libcloud.groovy

```diff
@@ -46,10 +48,92 @@ def replicate_to_clouds(pipecfg, basearch, buildID, stream) {
+            def c = config
+            // In 4.19+ we switched the command to upload to clouds to
+            // `cosa imageupload-<cloud>`
+            // https://github.com/coreos/coreos-assembler/pull/4074
+            def image_upload_cmd = "imageupload"
+
+            // fetch the aws vmdk image from s3 because ore upload still requires it in the
+            // builds directory. We're already working in an initialized cosa working
+            // directory, so we only need to perform the buildfetch and decompress here.
+            def s3_stream_dir = pipeutils.get_s3_streams_dir(pipecfg, stream)
+            shwrap("""
+                cosa buildfetch --build=${buildID} \
+                --arch=x86_64 --artifact=aws --url=s3://${s3_stream_dir}/builds \
+                --aws-config-file \${AWS_CONFIG_FILE} \
+                && cosa decompress --build=${buildID} --artifact=aws
+            """)
+
+            // Discover the latest Windows Server AMI to use as the winli-builder instance.
+            // The AMI rotates frequently with Windows updates and is not persistant in AWS
+            // for very long, so we need to find the most recent AMI ID.
+            // Windows Server 2022 was selected here because the Windows Server 2025 AMI does
+            // not allow legacy bios. WS2022 has an EOL date of 2026-10-13.
+            // https://learn.microsoft.com/en-us/lifecycle/products/windows-server-2022
+            // If WS2022 becomes inaccessible in the future and we still need BIOS for our
+            // winli images then we can just use one of our own previously created winli
+            // images and pass that to `--windows-ami` below.
+            def windows_server_ami_name = ""
+            windows_server_ami_name = "Windows_Server-2022-English-Full-Base-*"
+            def windows_server_ami = shwrapCapture("""
+                aws ec2 describe-images   \
+                    --region=${c.primary_region}   \
+                    --owners="amazon"   \
+                    --filters="Name=name,Values=${windows_server_ami_name}"   \
+                    --query="sort_by(Images, &CreationDate)[-1].ImageId"   \
+                    --output text
+            """)
+
+            // validate that we did actually get an AMI ID returned.
+            if (!(windows_server_ami_name != /^ami-[0-9][a-f]{17}$/)) {
+                error("Invalid Windows Server AMI ID: ${windows_server_ami}")
+            }
+
+            def extraArgs = []
+            if (c.grant_users) {
+                extraArgs += c.grant_users.collect{"--grant-user=${it}"}
+                extraArgs += c.grant_users.collect{"--grant-user-snapshot=${it}"}
+            }
+            if (c.tags) {
+                extraArgs += c.tags.collect { "--tags=${it}" }
+            }
+            if (c.public) {
+                extraArgs += "--public"
+            }
+            shwrap("""
+                cosa ${image_upload_cmd}-aws \
```

Comment:
Since we don't need to worry about backwards compat here we can juts go ahead and:
```suggestion
                cosa imageupload-aws \
```